accepted/cranelift-isel-isle-peepmatic.md

```diff
@@ -0,0 +1,926 @@
+complex because its design is intertwined with a particular backend
+design or strategy, or other incidental details, then one will not be
+able to use it effectively without understanding implementation
+details. On the other hand, if we have a clean rewrite-based
+semantics, one can easily convince oneself that a new rule is safe to
+add as long as it denotes a true equivalence. This is similar to how
+"equational reasoning", in functional programming terms (i.e., that
+one can always substitute a function invocation with its definition),
+makes a program's semantics easier to understand.
+
+The order in which rewrite rules will be applied can be influenced by
+user-specified priorities. However, the language definition is careful
+to *not* specify which rule must be applied first; priorities are just
+heuristics, or hints. This leaves room for the DSL use-case to apply
+rules differently: for example, we might test/fuzz a backend by
+applying rules in a different order. One should also be free to tweak
+priorities to attempt to optimize, without fear of breaking the
+semantics.
+    
+Note that the ordering is only undefined when there are multiple legal
+choices according to the types and available rules. In other words,
+this does *not* mean that we will arbitrarily steer into a dead-end in
+the graph of rewrite steps; an instruction that is lowerable should be
+lowerable with any rule ordering heuristic.
+
+# Extractors: Programmable Matching on Virtual Nodes
+
+The first key innovation that this DSL introduces is in how it allows
+patterns to *match* on input terms.
+
+A vanilla term-rewriting system operates on trees of nodes and
+transforms a single input tree (which is usually reified as an actual
+in-memory data structure) into a single output tree. While this is an
+easily-understood model, the requirement to define a *single* view of
+the input can unnecessarily restrict expressivity. Said another way:
+we sometimes want to use terms to represent certain properties of a
+value, and it is difficult to arrange the terms into a tree and then
+write the pattern-matching rules in a way that is flexible enough to
+support multiple types of queries.
+
+## Motivation
+
+For a concrete example, let us assume we a term type representing a
+constant integer value
+
+```plain
+    (iconst 42)
+```
+
+and we are performing instruction selection to lower to a CPU
+instruction set that supports multiple different *integer immediate
+formats* for its instructions. For example, AArch64 supports a 12-bit
+immediate (unshifted or shifted left 12 bits) in some of its
+arithmetic instructions, and a special "logic-instruction immediate"
+with bit-pattern generation in other instructions. In a conventional
+term-rewriting system, with helper functions defined appropriately, we
+might have rules like
+
+```plain
+    ;; 12-bit immediate form.
+    (rule (iconst val)
+          (when (is-aarch64-imm12-value val))  ;; Conditional predicate
```

Comment:
Might be nice to infer this using "typechecker" that takes constraints from the output instructions and propagates them to the input.

Reply:
That feels a *bit* too magic/implicit, but what can be done instead is that constructors that build the final instructions take strongly-typed values that can only be produced by extractors that check the relevant conditions. So the plan is to have e.g. an `Aarch64Imm12` value type, and this can only represent legal values. Similar type-encoded conditions can be used for other properties.

Reply:
We may also want to use this for things that are not typecheckable, such as a processor extension predicate.

```
(rule (iadd (i8x16 a) (i8x16 b))
      (when (has-3dnow))
      (3dnow-add a b))
```